### PR TITLE
fix(codex): 保留策略中断的错误提示

### PR DESCRIPTION
**Pull request by dashhuang**
### 摘要

Codex 的 Host Shell 策略拒绝已经开始的命令时，会请求 app-server 中断当前 turn。此前后续的 `turnCompleted(interrupted)` 会被当成用户点击 Stop，导致刚显示的策略错误被取消完成事件静默清除。

本 PR 为策略中断保留独立来源：策略命中后先发布非终态错误并请求中断，但 `turn/interrupt` 的 ACK 只视为“请求已收到”，不会提前将任务置为完成。只有真实 `turnCompleted` 到达后才结束本地 running 状态；若完成状态为 `interrupted`，则发布 terminal policy error 和 Codex 所需的 idle status，同时避免生成会清除错误的 `done(cancelled: true)`。

如果两次中断 ACK 都失败，marker 只保留来源、不屏蔽后续输出：真实结果为 `completed` / `failed` 时按 provider 结果正常收口；真实结果为 `interrupted` 时仍能识别为本次策略中断。若同一 turn 又启动另一条被策略拒绝的命令，会重新执行有界中断，而不是被旧 marker 跳过。用户主动 Stop 会覆盖同一 turn 上尚未结算的策略来源；Stop 后迟到的 blocked item 与策略 ACK 都不能重新把用户取消归因为策略错误，最终仍保持既有的静默 done(cancelled: true) 行为。

本轮 review 还补齐两条对称路径：策略中断 Plan turn 时立即结束当前 Plan cycle，避免下一条普通消息继续携带 Plan Mode；持久化的 `host-shell-command-blocked` 错误使用稳定 i18n key，并补齐五种支持语言。

### 变更类型

- [ ] `feat` 新功能
- [x] `fix` 缺陷修复
- [ ] `refactor` / `perf` 重构或性能优化
- [ ] `docs` / `test` / `chore` 文档、测试或工程维护
- [ ] 其他：

### 范围

- 关联 Issue / 需求：Codex 内部策略中断与用户 Stop 无法区分，错误提示被静默清除
- 本 PR 包含：Codex Host Shell 策略中断来源、ACK / completion 竞态收口、ACK 丢失回退、同 turn 重试、Plan cycle 清理、持久化错误五语本地化；对应事件顺序与恢复路径回归测试
- 明确不包含：Shell 命令策略解析器的误判修复；Claude Code / Pi 中断来源模型重构
- 用户可见变化：策略中断不会因 ACK 提前显示为完成；错误提示会在真实中断后保留并按界面语言显示；被策略中断的 Plan turn 不会污染下一条普通消息
- 是否存在 breaking change：无

### UI 变化

仅修改现有错误横幅/历史错误卡使用的五语文案；未修改组件、布局、颜色、动效或交互，Light / Dark 继续复用同一现有错误展示。

- 引用的设计规范：`DESIGN.md` §11.1 “Errors = what happened + what to do”——文案说明 Shell 命令被阻止，并明确下一步应使用内嵌 iOS Simulator 工具。

### 自动验证

```text
pnpm --filter @cindy/maker-core exec vitest run src/agents/codex/index.test.ts
结果：455 tests passed

pnpm check:i18n
结果：5 种语言 7190 个 key 一致

pnpm check:i18n-glossary
结果：无新增术语违规

pnpm --filter desktop run --if-present typecheck
结果：通过

pnpm --filter @cindy/maker-core run --if-present typecheck
结果：通过；该 package 当前未定义 typecheck script，按仓库门禁规则跳过

bash /Users/dash/Code/XD/dash/Skills/git/scripts/run-unit-gate.sh /Users/dash/Code/Cindy/cindy-fix-interrupt-origin
结果：GATE_EXIT=0；最新 HEAD source 在运行期间未变化

pnpm check:dco
结果：7 commits signed off

git diff --check
结果：通过
```

### 手工验证

不涉及：本次修复由 Codex 事件流、Desktop 错误恢复与 i18n 回归测试覆盖，未启动 Desktop 实机。

### 未执行的验证

- 未进行 Desktop 实机目检；未修改错误展示结构，文案复用现有 i18n / error 渲染路径。
- 额外尝试的 `pnpm --filter @cindy/maker-core build` 被仓库现有测试类型错误拦住；错误覆盖 Claude Code / Pi / Codex 旧测试位置，本次新增测试行未出现在错误清单。提交门禁使用仓库规定的 package typecheck 入口与全仓 `test:unit`，两者通过。

### 风险分类

- [ ] 无已知风险
- [ ] SQLite / migration
- [ ] system prompt
- [ ] 协议兼容
- [x] 权限 / 安全 / 用户数据
- [ ] 存量插件兼容（批准状态 / 指纹 / manifest 校验 / 安装布局 / 包格式）
- [ ] 原生层 / fingerprint / OTA
- [ ] 跨平台差异
- [ ] 其他：

### 影响与回滚

- 影响范围：仅 Codex 已开始执行的 Shell 命令被 Host 策略拒绝时，中断请求与真实 turn completion 之间的低频事件顺序，以及该错误的持久化本地化；策略判断和拒绝边界不变。
- 回滚 / 降级方式：回退本 PR，恢复原先非终态错误 + interrupt 的收口方式。
- Maker Core 指标评估：未修改 prompt、工具定义、模型路由或用量计算，缓存率不受影响。回归测试覆盖 ACK 成功但 completion 未到、completion 早于 ACK、ACK 两次失败后的正常完成、ACK 丢失后的 interrupted、同一 turn 再次命中策略时重新中断，Plan turn 被策略中断后的下一条普通消息，以及 policy hit → user Stop → late blocked item → interrupted completion 的来源覆盖顺序。
- 存量插件影响：无。

### 第 5 轮收敛检查点

- 不变量：每个 turn 在 turnCompleted 前只有一个权威 interrupt origin；显式用户 Stop 覆盖 host-policy，迟到 item / ACK 不得反向改写来源；只有来源仍为 host-policy 且权威状态为 interrupted 才发布 terminal policy error。
- 唯一 owner：turnInterruptOrigins[turnId] 持有 host-policy（blocked item、原因、ACK 状态）或 user-stop。policy itemStarted 建立或替换 host-policy；handle.abort 原子覆盖为 user-stop；turnCompleted 是唯一消费和删除入口；旧异步 ACK 只能在 identity 仍匹配时更新。
- 状态矩阵：host-policy pending + 同 item → 去重；host-policy 任意 ACK 状态 + 新 item → 新中断；user-stop + 任意迟到 blocked item / ACK → 保持 user-stop；completed / failed → provider 权威收口；interrupted + host-policy → terminal policy error + idle；interrupted + user-stop → done(cancelled: true)；重复 / 迟到 completion → 既有墓碑去重。
- 对称路径审计：覆盖 interrupt ACK 成功 / 失败 / 迟到，completion 早于 ACK，completed / failed / interrupted，用户 Stop 覆盖，Stop 后迟到 blocked item，以及同一 blocked item 重复 / 新 blocked item 继续出现。
- 可鉴别测试：同一 item 在 ACK 前重复只产生 1 次 TurnInterrupt；ACK 后新 blocked item 使调用数增至 2；policy hit 后用户 Stop、再收到 late blocked item 时仍只产生用户取消 done，旧的“只删除 marker”修法会在最后一个断言失败。

### 提交前检查

- [x] 已 review 完整 diff
- [x] 每个 commit 都带 DCO 签名（`git commit -s`，见 [DCO](../DCO)）
- [x] UI 改动已在「UI 变化」注明引用的设计规范
- [x] 未提交凭证、令牌或授权文件
- [x] 已补充必要文档（不需要额外文档）
- [x] 已确认测试结果或说明未执行原因

### CODE REVIEW

**Comment by MagicLizi** [comment]
@dashhuang 👋 这个 PR 还有 2 条 review conversation 没 resolve(packages/maker-core/src/agents/codex/index.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: packages/maker-core/src/agents/codex/index.ts [hunk 1]
```diff
@@ -8971,14 +8971,23 @@ export class CodexAgent extends BaseAgent {
...
+            // durable error before interrupting and tombstone the turn so the later
+            // turn/completed(interrupted) only performs bookkeeping instead of emitting
+            // done(cancelled:true), which would clear the policy error in Desktop.
+            terminalErroredTurnIds.add(params.turnId);
             eventQueue.push({
               type: 'error',
-              data: { message: hostPolicy.reason, isTerminal: false },
+              data: {
+                message: hostPolicy.reason,
+                isTerminal: true,
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply packages/maker-core/src/agents/codex/index.ts]: 已在 d17d8a465 修复：策略中断只有在 interrupt ACK 成功或已收到 turn 终态后才发布 terminal error，并紧跟 status(isRunning:false) 作为 Codex idle tail；随后 provider completion 只做 bookkeeping。新增回归测试锁定 error → idle 顺序及迟到 completion 不重复终态。

**Comment by MagicLizi** [comment]
@dashhuang 👋 这个 PR 还有 1 条 review conversation 没 resolve(packages/maker-core/src/agents/codex/index.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: packages/maker-core/src/agents/codex/index.ts [hunk 2]
```diff
@@ -8967,18 +8981,70 @@ export class CodexAgent extends BaseAgent {
...
+                  isTerminal: true,
+                  reason: 'host-shell-command-blocked',
+                },
+                source: 'codex',
+              });
+              handleTurnCompleted(
+                completedWhileInterrupting ?? ({
+                  threadId,
+                  turn: { id: params.turnId, status: 'interrupted' },
+                } as TurnCompletedParams),
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply packages/maker-core/src/agents/codex/index.ts]: 已在 ae93ed410 修复：turn/interrupt ACK 只表示请求被接收，不再合成本地 completion 或提前放开下一轮；任务保持 running，直到真实 turnCompleted 到达。ACK 丢失时保留不屏蔽输出的策略来源 marker，后续 interrupted 仍按策略中断收口，completed / failed 按 provider 结果正常收口；同一 turn 再次命中策略会重新请求中断。新增交错回归测试，Codex 定向 451/451、全仓单测门禁通过。

File: packages/maker-core/src/agents/codex/index.ts [hunk 3]
```diff
@@ -7680,6 +7689,44 @@ export class CodexAgent extends BaseAgent {
...
+      if (pendingHostPolicyInterrupt) {
+        pendingHostPolicyInterrupts.delete(turn.id);
+        if (terminalErroredTurnIds.has(turn.id) || completedTurnIds.has(turn.id)) {
+          handleTurnCompleted(params);
+          return;
+        }
+        if (
+          pendingHostPolicyInterrupt.acknowledgement === 'failed'
+          && turn.status !== 'interrupted'
+        ) {
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply packages/maker-core/src/agents/codex/index.ts]: 已在 85d722e19 修复：ACK 状态现在只控制同一 turn 是否允许重试中断，不再参与终态裁决。权威 turnCompleted 为 completed / failed 时始终按 provider 结果正常收口；只有真实 interrupted 才发布 terminal 策略错误。交错测试扩展为 interrupted / completed / failed 三种状态均早于 ACK，Codex 定向 453/453、全仓 test:unit 通过。

File: packages/maker-core/src/agents/codex/index.ts [hunk 4]
```diff
@@ -7680,6 +7689,42 @@ export class CodexAgent extends BaseAgent {
...
+          // interrupt. The provider's completed/failed status is authoritative
+          // even when it arrives before that ACK settles.
+          handleTurnCompleted(params);
+          return;
+        }
+        // The provider completion is the first proof that the denied command is
+        // actually finished. Close the turn with the policy error and Codex's
+        // required idle tail; the tombstone keeps interrupted from becoming a
+        // user-Stop done(cancelled:true).
+        terminalErroredTurnIds.add(turn.id);
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply packages/maker-core/src/agents/codex/index.ts:7719]: 已在 9c826ebd4 修复：策略中断收到真实 turnCompleted(interrupted) 后，会在终态墓碑抑制普通 completion UI 前清空 proposedPlanText、planCycleActive 与 currentTurnPlanModeActive；下一条普通消息恢复 collaborationMode: default。新增回归测试覆盖 Plan turn 被策略中断后继续发送普通消息。

File: packages/maker-core/src/agents/codex/index.ts [hunk 5]
```diff
@@ -7680,6 +7689,42 @@ export class CodexAgent extends BaseAgent {
...
+        eventQueue.push({
+          type: 'error',
+          data: {
+            message: pendingHostPolicyInterrupt.reason,
+            isTerminal: true,
+            reason: 'host-shell-command-blocked',
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply packages/maker-core/src/agents/codex/index.ts:7725]: 已在 9c826ebd4 修复：host-shell-command-blocked 现在稳定映射到 logic.errors.hostShellCommandBlocked，并补齐 en / zh-CN / zh-TW / ja / ko 五语文案；历史错误恢复测试锁定该 reason 映射。pnpm check:i18n 与 pnpm check:i18n-glossary 均通过。

**Comment by MagicLizi** [comment]
@dashhuang 👋 这个 PR 还有 2 条 review conversation 没 resolve(packages/maker-core/src/agents/codex/index.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: packages/maker-core/src/agents/codex/index.ts [hunk 6]
```diff
@@ -8967,18 +9020,47 @@ export class CodexAgent extends BaseAgent {
           });
           if (hostPolicy?.decision === 'deny') {
             discardPendingSpawnLineageIds(reservedChildThreadIds);
+            const existingPolicyInterrupt = pendingHostPolicyInterrupts.get(params.turnId);
+            if (existingPolicyInterrupt?.acknowledgement === 'pending') return;
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply packages/maker-core/src/agents/codex/index.ts]: 已在 0628f68e6 修复：策略中断 marker 现在记录 blocked itemId，只有同一 item 的 interrupt RPC 仍在飞时才去重；不同 item 是 turn 仍在执行的新证据，即使前一次尚未 ACK 也会覆盖 owner 并重新发起有界中断。ACK 成功会转为 acknowledged，失败转为 failed，权威 turnCompleted 统一删除 marker。交错测试同时锁定“pending 同 item 只发一次”和“ACK 后新 item 再次中断”，Codex 定向 454/454、全仓 test:unit 通过。

[the remark this answers is not in the document]
Comment (dashhuang) [reply packages/maker-core/src/agents/codex/index.ts]: 已在 0628f68e6 修复：去重判据已收敛为“同一 blocked item 且 interrupt RPC 仍 pending”；ACK 后 marker 转为 acknowledged，后续新 blocked item 会重新发起 interrupt，不再被旧 pending 状态拦截。回归测试覆盖 ACK 成功后、真实 completion 前出现第二条 blocked command，并确认 TurnInterrupt 调用从 1 增至 2；全仓 test:unit 通过。

File: packages/maker-core/src/agents/codex/index.ts [hunk 7]
```diff
@@ -7680,6 +7693,50 @@ export class CodexAgent extends BaseAgent {
 
     function handleTurnCompleted(params: TurnCompletedParams): void {
       const turn = params.turn;
+      const pendingHostPolicyInterrupt = pendingHostPolicyInterrupts.get(turn.id);
+      if (pendingHostPolicyInterrupt) {
+        pendingHostPolicyInterrupts.delete(turn.id);
```
[the remark this answers is not in the document]
Comment (Shinku-Chen) [reply packages/maker-core/src/agents/codex/index.ts]: 同意这个 P2 建议。用户 Stop 时应清除策略中断标记，避免将用户取消错误归因为策略错误。当前改动已解决主要问题，此优化建议在后续 PR 中完善。

**Comment by MagicLizi** [comment]
@dashhuang 👋 这个 PR 还有 1 条 review conversation 没 resolve(packages/maker-core/src/agents/codex/index.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: packages/maker-core/src/agents/codex/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (dashhuang) [reply packages/maker-core/src/agents/codex/index.ts]: 已在 8059a5f78 修复：每个 turn 现在由统一的 interrupt origin 记录 `host-policy | user-stop`。显式 `handle.abort()` 会原子覆盖策略来源，权威 `turnCompleted` 统一消费并清理；Stop 后迟到的 blocked item 不能再把用户取消改回策略错误，迟到的策略 ACK 也因 identity 不匹配被忽略。新增交错测试锁定 policy hit → user Stop → late blocked item → interrupted completion，最终只产生 `done(cancelled: true)`。Codex 测试 455/455、全仓 test:unit 门禁通过。

**Review by MagicLizi — APPROVED** [review]
✅ Auto review: no P0/P1 findings. Fix correctly introduces per-turn interruptOrigin tracking with proper ack/retry/failure handling, preventing misattribution of policy kills as user stops. Comprehensive test coverage across all edge cases.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by MagicLizi** [comment]
🧹 已合并！这次 interruptOrigin 追踪的设计很精巧——以 turn/completed 作为权威门、pending/acknowledged/failed 三态处理异步竞态、user-stop 不可被迟到的 policy 通知覆盖，每个边界都考虑到了。测试覆盖也全面，赞！
